core/src/test/java/org/apache/calcite/test/TopDownOptTest.java

```diff
@@ -65,92 +70,177 @@
  * </ol>
  */
 class TopDownOptTest extends RelOptTestBase {
-
-  protected DiffRepository getDiffRepos() {
-    return DiffRepository.lookup(TopDownOptTest.class);
-  }
-
-  Sql sql(String sql) {
-    VolcanoPlanner planner = new VolcanoPlanner();
-    // Always use top-down optimization
-    planner.setTopDownOpt(true);
-    planner.addRelTraitDef(ConventionTraitDef.INSTANCE);
-    planner.addRelTraitDef(RelCollationTraitDef.INSTANCE);
-
-    RelOptUtil.registerDefaultRules(planner, false, false);
-
-    // Keep deterministic join order
-    planner.removeRule(JoinCommuteRule.INSTANCE);
-    planner.removeRule(JoinPushThroughJoinRule.LEFT);
-    planner.removeRule(JoinPushThroughJoinRule.RIGHT);
-
-    // Always use merge join
-    planner.removeRule(EnumerableRules.ENUMERABLE_JOIN_RULE);
-    planner.removeRule(EnumerableRules.ENUMERABLE_CALC_RULE);
-
-    // Always use sorted agg
-    planner.removeRule(EnumerableRules.ENUMERABLE_AGGREGATE_RULE);
-    planner.addRule(EnumerableRules.ENUMERABLE_SORTED_AGGREGATE_RULE);
-
-    Tester tester = createTester().withDecorrelation(true)
-        .withClusterFactory(cluster -> RelOptCluster.create(planner, cluster.getRexBuilder()));
-
-    return new Sql(tester, sql, null, planner,
-        ImmutableMap.of(), ImmutableList.of());
-  }
+  // Ordering enforcement should not be done by rules
+  private static final List<RelOptRule> SORT_PROJECT_TRANSPOSE_RULES =
+      ImmutableList.of(
+          SortProjectTransposeRule.INSTANCE,
+          ProjectSortTransposeRule.INSTANCE
+      );
+
+  // Remove to always use merge join
+  private static final List<RelOptRule> NON_MERGE_JOIN_RULES =
+      ImmutableList.of(
+          EnumerableRules.ENUMERABLE_JOIN_RULE,
+          EnumerableRules.ENUMERABLE_CALC_RULE
```

Comment:
Sorry, maybe ENUMERABLE_CALC_RULE is not needed. I found that it is not registered by default.